runtime/runtime_storage.go

```diff
@@ -211,41 +212,122 @@ func (s *interpreterRuntimeStorage) writeValue(
 //
 func (s *interpreterRuntimeStorage) writeCached() {
 
+	type writeItem struct {
+		storageKey storageKey
+		value      interpreter.Value
+	}
+
+	var items []writeItem
+
 	for fullKey, entry := range s.cache {
 
 		if !entry.mustWrite && entry.value != nil && !entry.value.Modified() {
 			continue
 		}
 
+		items = append(items, writeItem{
+			storageKey: fullKey,
+			value:      entry.value,
+		})
+	}
+
+	// Don't use a for-range loop, as keys are added while iterating
+	for len(items) > 0 {
+		var item writeItem
+		item, items = items[0], items[1:]
+
 		var newData []byte
-		if entry.value != nil {
+		if item.value != nil {
+			var deferrals *interpreter.EncodingDeferrals
 			var err error
-			reportMetric(
-				func() {
-					newData, err = interpreter.EncodeValue(entry.value)
-				},
-				s.runtimeInterface,
-				func(metrics Metrics, duration time.Duration) {
-					metrics.ValueEncoded(duration)
-				},
-			)
+			newData, deferrals, err = s.encodeValue(item.value, item.storageKey.key)
 			if err != nil {
 				panic(err)
 			}
+
+			for deferredKey, deferredValue := range deferrals.Values {
+
+				deferredStorageKey := storageKey{
+					address: item.storageKey.address,
+					key:     deferredKey,
+				}
+
+				if !deferredValue.Modified() {
+					continue
+				}
+
+				items = append(items, writeItem{
+					storageKey: deferredStorageKey,
+					value:      deferredValue,
+				})
+			}
+
+			for _, deferralMove := range deferrals.Moves {
+
+				s.move(
+					deferralMove.DeferredOwner,
+					deferralMove.DeferredStorageKey,
+					deferralMove.NewOwner,
+					deferralMove.NewStorageKey,
+				)
+			}
 		}
 
 		var err error
 		wrapPanic(func() {
 			// TODO: fix controller
-			err = s.runtimeInterface.SetValue(
-				[]byte(fullKey.storageIdentifier),
-				[]byte{},
-				[]byte(fullKey.key),
-				newData,
-			)
+			err := s.runtimeInterface.SetValue(
+        item.storageKey.address[:],
+        nil,
+        []byte(item.storageKey.key),
+        newData,
+      )
 		})
 		if err != nil {
 			panic(err)
 		}
 	}
 }
+
+func (s *interpreterRuntimeStorage) encodeValue(
+	value interpreter.Value,
+	path string,
+) (
+	data []byte,
+	deferrals *interpreter.EncodingDeferrals,
+	err error,
+) {
+	reportMetric(
+		func() {
+			data, deferrals, err = interpreter.EncodeValue(value, []string{path}, true)
+		},
+		s.runtimeInterface,
+		func(metrics Metrics, duration time.Duration) {
+			metrics.ValueEncoded(duration)
+		},
+	)
+	return
+}
+
+func (s *interpreterRuntimeStorage) move(
+	oldOwner common.Address, oldKey string,
+	newOwner common.Address, newKey string,
+) {
+	// TODO: fix controller
```

Comment:
Unsure what this comment means, which controller is this referring to?

Reply:
This has been unfortunately cargo-culted for a while:

In the early design of the storage architecture there was the idea that fir each piece of stored data there is a "controller" and an "owner": 
The owner would "own" the data in their account and pay storage rent for it. The controller would have read/write permissions to the data.
For example with a non-fungible token, the user would store it, i.e. be the owner, but the controller would be able to modify it, e.g. the account which has the contract for the NFT, e.g. Dapper Labs has the CK contract, and kitties are stored in user accounts, but only the CK contract can change them.

We started with this notion in the storage API (`runtime.Interface`'s `GetValue` and `SetValue` functions have a `controller` and `owner` parameter, but we only use one at the moment.

After some discussion we realized that we're actually enforcing the ownership and control through the statically through the type system, so there is really no need to the separation in the storage API. We should remove/replace it with a single key at some point soon.